File: backend/src/test/kotlin/com/skewwhiffy/notation/model/note/AccidentalTest.kt
package com.skewwhiffy.notation.model.note

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class AccidentalTest {
  @Test
  fun `has correct abc when natural`() {
    val expected = ""
    val natural = Accidental.natural

    val actual = natural.abc

    assertThat(actual).isEqualTo(expected)
  }

  @Test
  fun `has correct abc when flat`() {
    val expected = "_"
    val flat = Accidental.flat

    val actual = flat.abc

    assertThat(actual).isEqualTo(expected)
  }

  @Test
  fun `has correct abc when sharp`() {
    val expected = "^"
    val sharp = Accidental.sharp

    val actual = sharp.abc

    assertThat(actual).isEqualTo(expected)
  }

  @Test
  fun `displays correctly when natural`() {
    val expected = ""
    val natural = Accidental.natural

    val actual = natural.displayString

    assertThat(actual).isEqualTo(expected)
  }

  @Test
  fun `is flat when flattening natural`() {
    val natural = Accidental.natural

    val actual = natural.flat

    assertThat(actual).isEqualTo(Accidental.flat)
  }

  @Test
  fun`be sharp when sharpening natural` () {
    val natural = Accidental.natural

    val actual = natural.sharp

    assert(actual == Accidental.sharp)
  }

  fun`display flat when flat" () {
    val expected = "b"
    val flat = Accidental.flat

    val actual = flat.displayString

    assert(actual == expected)
  }

  fun`display double flat when flattening flat" () {
    val expected = "bb"
    val flat = Accidental.flat

    val actual = flat.flat

    assert(actual.displayString == expected)
  }

  fun`be natural when sharpening flat" () {
    val flat = Accidental.flat

    val actual = flat.sharp

    assert(actual == Accidental.natural)
  }

  List(3, 7).foreach(flats => {
    it should s"display correctly when $flats flats" () {
      val expected = "b".repeat(flats)
      val accidental = List.range(0, flats).foldRight(Accidental.natural)((_, acc) => acc.flat)

      val actual = accidental.displayString

      assert(actual == expected)
    }
  })

  fun`display sharp when sharp" () {
    val expected = "#"
    val sharp = Accidental.sharp

    val actual = sharp.displayString

    assert(actual == expected)
  }

  fun`display double sharp when sharpening sharp` () {
    val expected = "x"
    val sharp = Accidental.sharp

    val actual = sharp.sharp

    assert(actual.displayString == expected)
  }

  fun`be natural when flattening sharp` () {
    val sharp = Accidental.sharp

    val actual = sharp.flat

    assert(actual == Accidental.natural)
  }

  List(6, 10).foreach(numberOfSharps => {
    it should s"display correctly when $numberOfSharps" in {
      val expected = "x".repeat(numberOfSharps / 2)
      val accidental = List.range(0, numberOfSharps).foldRight(Accidental.natural)((_, it) => it.sharp)

      val actual = accidental.displayString

      assert(actual == expected)
    }
  })

  List(7, 13).foreach(numberOfSharps => {
    it should s"display correctly when $numberOfSharps" in {
      val expected = "x".repeat(numberOfSharps / 2) + "#"
      val accidental = List.range(0, numberOfSharps).foldRight(Accidental.natural)((_, it) => it.sharp)

      val actual = accidental.displayString

      assert(actual == expected)
    }
  })

  fun`equate equivalent accidentals` in {
    def accidental = Accidental(5)

    val first = accidental
    val second = accidental

    assert(first == second)
  }
}
   */
}

File: backend/src/model/accidental.test.ts
import accidentalFactory from "./accidentalFactory";
import chai from 'chai'

chai.should();

describe('accidental', () => {
    it('has correct abc when natural', () => {
        const natural = accidentalFactory.natural;

        const actual = natural.abc;

        actual.should.equal('')
    })

    it('has correct abc when flat', () => {
        const flat = accidentalFactory.flat;

        const actual = flat.abc;

        actual.should.equal('_');
    })

    it('has correct abc when sharp', () => {
        const sharp = accidentalFactory.sharp;

        const actual = sharp.abc;

        actual.should.equal('^');
    })

    it('displays correctly when natural', () => {
        const natural = accidentalFactory.natural;

        const actual = natural.displayString;

        actual.should.equal('');
    })

    it('when flattening natural then is flat', () => {
        const natural = accidentalFactory.natural;

        const actual = natural.flatten

        actual.equals(accidentalFactory.flat).should.be.true
    })

    it('when sharpening natural then is sharp', () => {
        const natural = accidentalFactory.natural;

        const actual = natural.sharpen

        actual.equals(accidentalFactory.sharp).should.be.true;
    })

    it('when flat then displays flat', () => {
        const flat = accidentalFactory.flat;

        const actual = flat.displayString;

        actual.should.equal('b');
    })

    it('when flattening flat then displays double flat', () => {
        const flat = accidentalFactory.flat;

        const actual = flat.flatten;

        actual.displayString.should.equal('bb');
    })

    it('when sharpening flat then is natural', () => {
        const flat = accidentalFactory.flat;

        const actual = flat.sharpen;

        actual.equals(accidentalFactory.natural).should.be.true;
    });


    [3, 7].forEach(flats => {
        it(`displays correctly with ${flats} flats`, () => {
            const expected = 'b'.repeat(flats);
            const accidental = [...new Array(flats)].reduce((prev) => prev.flatten, accidentalFactory.natural);

            const actual = accidental.displayString;

            actual.should.equal(expected);
        })
    })

    it('when sharp displays sharp', () => {
        const sharp = accidentalFactory.sharp;

        const actual = sharp.displayString;

        actual.should.equal('#');
    })

    it('when sharpening sharp then displays double sharp', () => {
        const sharp = accidentalFactory.sharp;

        const actual = sharp.sharpen.displayString;

        actual.should.equal('x');
    })

    it('when flattening sharp then is natural', () => {
        const sharp = accidentalFactory.sharp;

        const actual = sharp.flatten;

        actual.equals(accidentalFactory.natural).should.be.true;
    });

    [6, 10].forEach(sharps => {
        it(`displays sharps correctly for ${sharps} sharps`, () => {
            const expected = 'x'.repeat(sharps / 2);
            const accidental = [...new Array(sharps)].reduce(prev => prev.sharpen, accidentalFactory.natural);

            const actual = accidental.displayString;

            actual.should.equal(expected);
        })
    });

    [7, 13].forEach(sharps => {
        it(`displays sharps correctly for ${sharps} sharps`, () => {
            const expected = 'x'.repeat((sharps - 1) / 2) + '#';
            const accidental = [...new Array(sharps)].reduce(prev => prev.sharpen, accidentalFactory.natural);

            const actual = accidental.displayString;

            actual.should.equal(expected);
        })
    })
});